Rubberduck.VBEEditor/Class1.cs

```diff
@@ -0,0 +1,964 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.ComTypes;
+using Reflection = System.Reflection;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+// create some aliases as there are conflicts between these types in InteropServices and InteropServices.ComTypes
+using TYPEATTR = System.Runtime.InteropServices.ComTypes.TYPEATTR;
+using TYPEFLAGS = System.Runtime.InteropServices.ComTypes.TYPEFLAGS;
+using DISPPARAMS = System.Runtime.InteropServices.ComTypes.DISPPARAMS;
+using IMPLTYPEFLAGS = System.Runtime.InteropServices.ComTypes.IMPLTYPEFLAGS;
+using TYPEKIND = System.Runtime.InteropServices.ComTypes.TYPEKIND;
+using INVOKEKIND = System.Runtime.InteropServices.ComTypes.INVOKEKIND;
+using VARDESC = System.Runtime.InteropServices.ComTypes.VARDESC;
+using TYPELIBATTR = System.Runtime.InteropServices.ComTypes.TYPELIBATTR;
+using FUNCDESC = System.Runtime.InteropServices.ComTypes.FUNCDESC;
+using ELEMDESC = System.Runtime.InteropServices.ComTypes.ELEMDESC;
+using IVBE = Rubberduck.VBEditor.SafeComWrappers.Abstract.IVBE;
+
+// USAGE GUIDE:   see class VBETypeLibsAPI for demonstrations of usage.
+//
+// The root object for exposure of the type libraries is TypeLibsAccessor_VBE.  It takes the VBE in its construtor.
+// The main wrappers (TypeLibWrapper_VBE and TypeInfoWrapper_VBE) can be used as regular ITypeLib and ITypeInfo objects through casting.
+//
+// THIS IS A WORK IN PROGRESS.  ERROR HANDLING NEEDS WORK, AS DOES A FEW OF THE HELPER ROUTINES.
+//
+// WARNING: when using TypeLibsAccessor_VBE directly, do not cache it  
+//   The type library is LIVE information, so consider it a snapshot at that very moment when you are dealing with it
+//   Make sure you call TypeLibsAccessor_VBE.Dispose() as soon as you have done what you need to do with it.
+//   Once control returns back to the VBE, you must assume that all the ITypeLib/ITypeInfo pointers are now invalid.
+//
+// CURRENT LIMITATIONS:
+// At the moment, enums and UDTs are not exposed through the type libraries
+// Constants names are not available
+
+// IMPLEMENTATION DETAILS:
+// There are two significant bugs in the VBE typeinfos implementations that we have to work around.
+// 1)  Some implementations of ITypeInfo provided by the VBE will crash with an AV if you call IProvideClassInfo::GetClassInfo on them.
+//      And guess what method the CLR calls on all COM interop objects when creating a RCW?  You guessed it.
+//      So, we use an aggregation object, plus ITypeInfo and ITypeLib wrappers to circumvent this VBE bug.
```

Comment:
Just so I'm clear since this is a different interface, are we saying that implementing a `ITypeInfo` or `ITypeLib` implies that the same will implement `IProvideClassInfo`, which we omit when we aggregate?

Reply:
No, it's entirely optional to support IProvideClassInfo.  The VBE happens to implement it on ITypeInfo objects that it hosts, but there is a bug in the implementation.